### ISSUE TITLE
Add multivariate categorical outcome mode with variable cardinality

### ISSUE THREAD

**Issue by jarrod-dalton** [jarrod-dalton/multioutcome-cvae#1]
## Motivation

Some applications need to model several jointly distributed discrete outcomes
whose numbers of levels differ. Binary outcomes are the two-level special case,
while other outcomes may have three or more unordered or ordered levels.

One motivating example is a medication-regimen transition model containing:

- binary prescription indicators;
- antihypertensive class count with levels `0`, `1`, `2`, `3`, and `4+`;
- statin intensity with levels `none`, `low`, `moderate`, and `high`; and
- a categorical pulmonary-medication summary.

Fitting separate models would lose shared-latent dependence across these
outcomes. Treating one-hot columns as independent Bernoulli outcomes retains
some dependence but can generate invalid multiple-level or no-level results.

## Proposed Scope

Add `outcome_type="categorical"` for **multivariate categorical outcomes with
variable cardinality**. Every semantic outcome has at least two levels, so this
mode supports binary outcomes without requiring a separate Bernoulli component.

Keep the existing `bernoulli` and `gaussian` modes unchanged for backward
compatibility. This issue does not propose mixtures of categorical and
continuous/count outcomes.

Suggested schema:

```python
outcome_schema = [
    {
        "name": "antihypertensive_class_count",
        "levels": ["0", "1", "2", "3", "4+"],
    },
    {
        "name": "statin_intensity",
        "levels": ["none", "low", "moderate", "high"],
    },
    {
        "name": "metformin_prescribed",
        "levels": ["0", "1"],
    },
]
```

Level order is canonical metadata. The initial implementation may treat every
outcome as nominal; an ordinal likelihood is not required here.

## Suggested Data Contract

- `Y` remains a two-dimensional matrix with one integer-coded column per
  semantic outcome.
- Codes are zero-based positions in the corresponding `levels` vector.
- `Y_mask`, when supplied, has the same semantic shape as `Y`: one mask value
  per row and outcome, not one value per internal decoder logit.
- Generated samples return the same integer-coded matrix shape as `Y`.
- Fitted/exported artifacts retain the exact outcome names, level order, and
  decoder slices needed to recover labels safely.

### Network representation

- Derive `encoded_y_dim = sum(len(levels_j))` separately from the number of
  semantic outcomes.
- One-hot encode each observed categorical outcome before passing `Y` into the
  encoder.
- For a masked outcome, use an all-zero one-hot group for the encoder and omit
  that outcome from reconstruction loss.
- Reuse the existing shared decoder core and emit one flat logits tensor with
  `sum(len(levels_j))` columns.
- Maintain deterministic contiguous slices for each named outcome.

### Loss and generation

- Compute categorical cross-entropy once per observed semantic outcome.
- Sum or average across outcomes using an explicit, documented normalization;
  defaulting to equal weight per semantic outcome avoids weighting a variable
  merely because it has more levels.
- During generation, draw one shared latent `Z` per requested joint sample and
  sample one `Categorical(logits=...)` value from each outcome slice.
- This guarantees one valid level per outcome while preserving dependence
  induced through the shared latent representation.

### Prediction API

- Return named per-outcome probability matrices, since different outcomes have
  different numbers of levels. For example:

```python
{
    "probabilities": {
        "antihypertensive_class_count": array_of_shape_n_by_5,
        "statin_intensity": array_of_shape_n_by_4,
        "metformin_prescribed": array_of_shape_n_by_2,
    }
}
```

- `predict_mean()` should not silently average integer category codes. It can
  either be unsupported for categorical models or require an explicit mapping
  from levels to numeric values.
- Keep `predict_proba()` behavior for existing Bernoulli models unchanged.

### R export

- Introduce a new export format version carrying `outcome_schema` and exact
  decoder slices.
- Generate base-R helpers that return named per-outcome probability matrices
  and sample exactly one level per outcome using a shared latent draw.
- Preserve existing Bernoulli export behavior and tests.

## Validation and Tests

- Reject duplicate outcome names, duplicate levels, fewer than two levels,
  non-integer codes, and codes outside the declared range.
- Verify two-level outcomes train and generate successfully in categorical mode.
- Verify every generated value is a valid code for its outcome.
- Verify masks exclude only the intended semantic outcomes from loss.
- Verify inference batching does not change output shapes or level ordering.
- Add an end-to-end simulation demonstrating recovery of marginal level
  frequencies and cross-outcome dependence.
- Add Python save/load and base-R export parity tests with fixed latent draws.
- Confirm existing Bernoulli and Gaussian tests remain unchanged and passing.

## Explicit Non-Goals

- Mixed categorical and Gaussian/Poisson outcomes in one model.
- Cumulative-link or other ordinal-specific likelihoods.
- Automatic clinical interpretation of level order.
- Changes to existing Bernoulli artifact or public API behavior.

## Acceptance Criteria

- A single categorical CVAE can fit and generate a joint outcome vector whose
  variables have different numbers of levels, including binary variables.
- Generated vectors always contain exactly one valid level per outcome.
- Shared latent draws continue to induce cross-outcome dependence.
- Outcome names and level order survive Python persistence and base-R export.
- Documentation includes one mixed-cardinality example and migration guidance
  explaining when to use `categorical` versus the existing `bernoulli` mode.

**Comment by jarrod-dalton** [comment]
Implemented on `main` in commit 8e1208c.

The categorical engineering contract now includes mixed cardinalities (including all-two-level schemas), semantic masking, grouped softmax loss, named probability outputs, shared-latent integer-code generation, bounded expanded batches, versioned Python checkpoints, and dependency-free base-R format-v2 export.

Release-gating CI passed for Python 3.9, 3.10, 3.11, and the R parity job: https://github.com/jarrod-dalton/multioutcome-cvae/actions/runs/33682854866

The simulation test confirms the scoped recovery criteria from this issue, but it does not establish general probabilistic predictive validity. That broader validation work will be tracked separately so closing this issue is not interpreted as a goodness-of-fit claim.

**Comment by jarrod-dalton** [comment]
Follow-up probabilistic predictive validation is tracked in #2: https://github.com/jarrod-dalton/multioutcome-cvae/issues/2
